### PR TITLE
Cleanup type information for ember-metal

### PR DESCRIPTION
**Pull request by mike-north**
This PR is a starting point for cleaning up some type information in ember-metal.

At a high level, here are some common things you'll find
- Call signature return types have been made explicit
    * Why? - [so that if changes are made to a function that deviate from the original developer intent, the type-checking will error at the definition of the function instead of all call-sites](https://agentcooper.github.io/typescript-play/#code/GYVwdgxgLglg9mABKSBGAFAQwE4HMBciYIAtgEYCm2AlIgN4BQiziECAzlEYgLyICymKAAsAdNkxgAJnBLpaAKkQ5cAbiYsA9AoUaWSgIJSpMMLkQiKiADamrla3ADuibBXYhrUdolN7mSlAAngAOFAC0EMIUEADWiFTYcNg+QhbRrJjW1ojsMFDuiHDAyOAQGNT+iAqa-pqaviXoSAA8iAAMtG5QINhIxNnqLK4UPX2IAAYAJHRgAL4T6nMMDCjQ8EhrAExYeITE5FTUhJzYpuaMw2xgnNx8giLikjJyisp4Q1o6VYbG5+lWWxgewURwuNweLw+PzDaoWUIRKIxeKJZKpLiWRBSCjAUz5Da5fJWYqlSA7SqwpQ-RAAVTA1ncPkxFAAHiFbBB8iMxkhgmFfD4opJcBQpIh0DBRBRRBY4LkoGczIgAD5ETzWCnDJRgOUQLLWcJ5Ao+JwwbKsWTszAwrW1Yb1Rri1odLqjXr9dWfZjdd2TGbzRYMZYrBjXW4ANyyqEwJwV-z4awwAGYk9R1GGuJHrFsY-LFeYE2Udim0ysM4gs6gyLH87xSeV0OES+mOJmsltq3n4-Wdk3U6ogA)
- Where possible (and simple), guard functions that returned `boolean` now appropriately narrow types
    * Why? - narrowing types in branching (`if`, `switch`, etc...) removes the need to explicitly cast
    * see [the User-Defined Type Guards section here](https://basarat.gitbooks.io/typescript/docs/types/typeGuard.html)
- At least since TS 2.5, any type union with `any` evaluates to an `any`
    * To avoid confusion, these are just changed to `any` where we can make very few assumptions (i.e., property access from an object)
    * In places where we know the value will always be truthy, we can use the `{}` type
    * In places where we know the vaue will never be truthy, we can define an `unusable` type
    * examples of these concepts and how they relate to `any` are [here](https://agentcooper.github.io/typescript-play/#code/C4TwDgpgBArgdjAzgQwEYBtoF4oPeqAH1jgBMIAzASzglKNwgDcIAnBpgeytIG4AofgHoAVCP5QRUACoALaAG8AvlFCQoyfJwDuiKBU7tkcEMFk0A5lE5m2q2cdwx8AGhLlqteoZIBrODpwEiJC-JjAUAAeAFxQylA4ACwCkQlQAOTyWukpaQoaFhCxAMwArFBKuTgAFACUCQB8Gcis0LaI0BABMBayUBCRyAC2YJh6IBDAAPw5-Kk48B40dFVO+Kvp6Rp68P6BG1vIerQsrAKCouKSMvIaJqrg0JroOnoGRiZmlsGh4VAgsWMIDSyX4wJwmQg2QE4PclGWfDBaTw6BhaU22z8AW0cDREMOx2YbHOl2CN2g8CQaEwD3Uz1e+h8QK+cCsZmQEQAxo4TnZUG1HgBaCAARxgVCYmi6EWAnA0qlYMFAUEl6BgEB+-Ao8E5wConDgjM4AEY6rFKSgMIolFqdXqDUaAExmkhUq1xKCtYAwVi4iq2uC6-WGgycYoui3UxSeyY+w2LeFeXj+7WB+0hzicRIRhCWmn5L1xtaoipAA)

CC: @krisselden

### CODE REVIEW

**Comment by rwjblue** [comment]
@mike-north - What are the TODO's here? It seems good at this point, is there any issue with landing these types of fixes incrementally?

**Comment by mike-north** [comment]
@rwjblue - @krisselden wanted to review this to ensure we weren't losing developer intent -- specifically around the types that union with `any` sometimes meaning "opaque" or "any" or `{}` in various situations

**Review by rwjblue — APPROVED** [review]
This LGTM, will leave open for a bit so @krisselden has a chance to review...

File: packages/@ember/-internals/metal/lib/descriptor.ts [hunk 1]
```diff
@@ -23,16 +23,16 @@ class NativeDescriptor extends Descriptor {
...
   }
 
-  setup(obj: object, key: string, meta: Meta) {
+  setup(obj: object, key: string, meta: Meta): void {
     Object.defineProperty(obj, key, this.desc);
     meta.writeDescriptors(key, this);
   }
 
-  get(obj: object, key: string): any | null | undefined {
+  get(obj: object, key: string): any {
```
[the remark this answers is not in the document]
Comment (mike-north) [reply packages/@ember/-internals/metal/lib/descriptor.ts]: wish we could, but we cannot due to

```ts
get(obj, "foo.bar.baz");
```

**Pull request merged**